src/backend/providers/playnite/PlayniteMetadataParser.cpp

```diff
@@ -0,0 +1,173 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include "PlayniteMetadataParser.h"
+
+#include "Log.h"
+#include "PlayniteComponents.h"
+#include "PlayniteJsonHelper.h"
+
+#include <QDirIterator>
+#include <QJsonDocument>
+#include <QJsonObject>
+
+namespace providers {
+namespace playnite {
+
+PlayniteMetadataParser::PlayniteMetadataParser(QString log_tag, const QDir& playnite_dir)
+    : m_log_tag(std::move(log_tag))
+    , m_playnite_dir(playnite_dir)
+    , dir_filters(QDir::Files | QDir::Readable | QDir::NoDotAndDotDot)
+    , json_ext_list(QStringList() << QStringLiteral("*.json"))
+{}
+
+PlayniteComponents PlayniteMetadataParser::parse_metdata() const
+{
+    PlayniteComponents components;
+    components.platforms = parse_platform_metadata();
+    components.sources = parse_source_metadata();
+    components.companies = parse_company_metadata();
+    components.genres = parse_genre_metadata();
+    components.emulators = parse_emulator_metadata();
+    components.games = parse_game_metadata();
+    return components;
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_platform_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/platforms"));
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_source_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/sources"));
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_company_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/companies"));
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_genre_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/genres"));
+}
+
+HashMap<QString, PlayniteEmulator> PlayniteMetadataParser::parse_emulator_metadata() const
+{
+    HashMap<QString, PlayniteEmulator> emulator_map;
+    QDirIterator dir_it(m_playnite_dir.filePath(QStringLiteral("library/emulators/")), json_ext_list, dir_filters);
+    while (dir_it.hasNext()) {
+        PlayniteEmulator emu;
+        const QString file_path = dir_it.next();
+        const QJsonObject json_object = get_json_object_from_file(file_path);
+        const auto emu_json = JsonObjectHelper(json_object);
+        emu.id = emu_json.get_string(QStringLiteral("Id"));
+        emu.name = emu_json.get_string(QStringLiteral("name"));
+        auto profiles = emu_json.get_json_array(QStringLiteral("Profiles"));
+        for (const auto profile : profiles) {
```

Comment:
You're possibly making a copy of each profile. You can use `const auto& profile`, just to make sure. Writing out the type for `profile` (and `profiles` above) would make it even more readable (they are just QJsonArray/Object/Value I guess?).